### ISSUE TITLE
fix(compute): mask paged-attention V-cache padding bytes so NaN cannot defeat score masking

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#13165]
# fix(compute): mask paged-attention V-cache padding bytes so NaN cannot defeat score masking

```routing
lane: compute
paths: ["internal/compute/cuda_block_table.go", "internal/compute/cuda_paged_attention.go", "internal/compute/paged_attention_nan_mask_test.go"]
expected_steps: 5
```

## Current state

FAK models paged attention as a logical-to-physical block table
(`internal/compute/cuda_block_table.go:44-54` `CUDABlockTable`, `:57`
`NewCUDABlockTable`, `:59` "standard PagedAttention default" of 16 tokens per
block). A logical block is allocated whole (`:68` `AppendLogicalTokens`) and its
`Tokens` counter (`:14`) tracks the *written* prefix — i.e. the slots beyond the
sequence boundary inside the last physical block are allocated but never written.
There is **no V-cache padding mask** anywhere in the package: a raw grep over
`internal/`/`pkg/` for a V-side padding-mask helper (`mask_v_cache`,
`vpadd`, `V*NaN*`) returns nothing.

vLLM fixes exactly this failure class in
`csrc/rocm/attention.cu:62-140`: unwritten paged-attention **V-cache** padding can
hold NaN (never-written device memory is not guaranteed zero), and `0 * NaN =
NaN`, so a *score-only* mask is defeated — the masked-out lane contributes NaN
through the value accumulation even though its softmax weight is zero. vLLM masks
the **V values by byte** at the `seq_len` boundary before accumulation
(`mask_v_cache_padding`, `mask_16b_v_cache_boundary`, `mask_8b_v_cache_boundary`)
and skips fully-invalid tail vectors; callers sit around `:844-960`. The
invariant: every V slot at index `>= seq_len` inside the final physical block must
be forced to a benign zero before it can enter the `sum(p * V)` accumulation.

## Value

- Centrality: Core — a correctness invariant on the paged-attention decode/prefill critical path.
- P1 Context: advanced - the failure is a silent NaN leak into logits, the most expensive class to debug downstream.
- P2 Net value: advanced - removes a whole silent-wrong-answer class for any partial-block (blockSize > written prefix) paged decode.
- P3 Adaptation: advanced - ports vLLM's byte-wise V-padding mask and tail-vector skip into fak's block table / paged-attention path.
- P4 Operations: preserved - local to `internal/compute`; no ABI change, no new public surface.

## Scope class

Deep - bounded kernel plus a falsifiable NaN-padding test.

## Why this is next

Paged attention on Strix Halo is the memory-bound decode spine and FAK already
allocates whole blocks with a partial written prefix (`cuda_block_table.go:68`).
The moment V is gathered from physical blocks whose tail slots were never written,
the score mask is no longer sufficient. The fix is cheap, local, and falsifiable,
and it must land before any wider speculative/ragged-block path multiplies the
number of partially-written physical blocks in flight.

## Working spine

Identify the last partial physical block -> mask V bytes at the `seq_len` boundary
-> skip fully-invalid tail vectors -> prove a NaN-poisoned padding slot cannot
change the `sum(p * V)` result -> parity against a clean-slot reference.

## Core through-line

1. Add a V-side padding mask helper (16-bit and 8-bit variants, mirroring
   `mask_16b_v_cache_boundary` / `mask_8b_v_cache_boundary`) operating at the
   `seq_len` boundary of the final physical block.
2. Wire it into the paged-attention value gather so the mask runs *before* the
   `sum(p * V)` accumulation.
3. Skip fully-invalid tail vectors rather than accumulating them as zero-weight
   NaN carriers.
4. Assert `0*NaN` can no longer leak: a V slot filled with NaN at `index >= seq_len`
   yields the same result as a clean zero slot.

## Gold-plating boundary

- Strictly excluded: changing the block size default (`cuda_block_table.go:59`) or
  the refcount/CoW discipline.
- Strictly excluded: altering the QK score mask itself — only the V-side padding
  path is in scope.
- Strictly excluded: any hardware/throughput claim; this is [SW-VERIFIED] only.

## Likely files

- `internal/compute/cuda_block_table.go`
- `internal/compute/cuda_paged_attention.go`
- `internal/compute/paged_attention_nan_mask_test.go`

If the paged-attention kernel lives under a differently named file, use the
closest real name and note the mapping in the resolving commit.

## Done condition

Paged-attention V-cache padding bytes past the `seq_len` boundary are masked to
zero before accumulation; a NaN-poisoned padding slot cannot alter the softmax-V
result.

## Witness

```bash
go -C ..\fak test ./internal/compute -run TestPagedAttentionVPaddingNaNMask -count=1
```

## Acceptance gate

`go -C ..\fak test ./internal/compute -run TestPagedAttentionVPaddingNaNMask -count=1` passes; `go -C ..\fak test ./internal/compute -count=1` remains green.

## Closure binding

Resolving commit cites this issue and carries conventional trailer (fak compute).

## Parent context

Native paged-attention / KV-MMU serving spine under epic #2236 and the paged-KV
serving baseline (#8395); unblocks safe partial-block and speculative decode
reuse (#12725).

## Definition of done

- [ ] 1. [SW-VERIFIED] A V-cache padding mask zeroes bytes at/after the `seq_len` boundary before the `sum(p * V)` accumulation.
- [ ] 2. [SW-VERIFIED] Fully-invalid tail vectors are skipped, not accumulated as zero-weight carriers.
- [ ] 3. [SW-VERIFIED] `TestPagedAttentionVPaddingNaNMask` shows a NaN-poisoned padding slot gives the same result as a clean-slot reference.
- [ ] 4. [SW-VERIFIED] `go -C ..\fak test ./internal/compute -count=1` passes.

## Work estimate
Estimate: 3 points. Uncertainty: operator-supplied estimate; revise through the parent denominator when evidence changes.

## Overall completion contribution
Contribution: 3/8 points.

## Completion standard
production

## Target operating envelope
- correctness: >= 1 unit test

## Witnessed operating envelope
- correctness: >= 1 unit test

**Comment by anthony-chaudhary** [comment]
Verification pass 2026-09-19 (burndown campaign #1838): NOT closed. The declared witness test internal/compute/paged_attention_nan_mask_test.go (TestPagedAttentionVPaddingNaNMask) does not exist; go test ./internal/compute -run TestPagedAttentionVPaddingNaNMask -count=1 exits 0 via [no tests to run] — a phantom witness. To close: author the declared V-padding NaN mask helper + test.
